Ensure that assets can be displayed in production

gsmendoza opened the pull request:
## Installing the gem locally

From your start_bootstrap_admin_theme directory:

1. Uninstall the gem: `gem uninstall start_bootstrap_admin_theme`
1. Build the gem: `gem build start_bootstrap_admin_theme`
1. Install the gem from the start_bootstrap_admin_theme directory: `gem install start_bootstrap_admin_theme`

## Scenario: Running the script should generate the asset files

Given I have downloaded and extracted [Start Bootstrap Admin Theme 2](https://startbootstrap.com/themes/sb-admin-2/)
When I run the install script against the downloaded folder e.g. 

```bash
./script/install/script.rb ~/Downloads/startbootstrap-sb-admin-2-gh-pages/
```

Then I should see the file updates in the "Rerun install script" commit

## Scenario: The gem doesn't have missing assets when ran in a development environment

Given I have installed the gem locally (see "Installing the gem locally")
And I have downloaded the [start_bootstrap_admin_theme_demo](https://github.com/SnackNation/start_bootstrap_admin_theme_demo) app
And my start_bootstrap_admin_theme and start_bootstrap_admin_theme_demo projects are using the same Ruby
And I have installed the bundle for the start_bootstrap_admin_theme_demo app
When I run the start_bootstrap_admin_theme_demo app in development environment
And I visit http://localhost:3000/start_bootstrap_admin_theme/pages/index.html
Then I should see all the assets (JS, stylesheets, images, fonts) are loaded

## Scenario: The gem doesn't have missing assets when ran in a production environment

Given I have installed the gem locally (see "Installing the gem locally")
And I have downloaded the start_bootstrap_admin_theme_demo app
And my start_bootstrap_admin_theme and start_bootstrap_admin_theme_demo projects are using the same Ruby
And I have installed the bundle for the start_bootstrap_admin_theme_demo app
And I have precompiled the assets i.e. `rails assets:precompile`
When I run the start_bootstrap_admin_theme_demo app in production environment
And I visit http://localhost:3000/start_bootstrap_admin_theme/pages/index.html
Then I should see all the assets (JS, stylesheets, images, fonts) are loaded

gsmendoza committed:
Fix: Add vendor to gem's files

gsmendoza committed:
[WIP] Regenerate engine

The commit is the unedited output of

```
rails plugin new start_bootstrap_admin_theme --mountable --skip-action-mailer --skip-action-mailbox --skip-action-text --skip-active-record --skip-active-storage --skip-action-cable --skip-turbolinks --skip-test --skip-system-test
```

gsmendoza committed:
Restore and fix APP_PATH

Point it to spec/dummy/config/application.

gsmendoza committed:
Restore app pre-update

gsmendoza committed:
Delete unnecessary Rails files

gsmendoza committed:
Precompile all vendored assets

gsmendoza committed:
Ensure that demo assets can be displayed in production

Use Rails asset URL helpers in HTML and Font Awesome files.

gsmendoza committed:
Rerun install script

gsmendoza commented on lib/start_bootstrap_admin_theme/engine.rb:
```diff
@@ -1,5 +1,15 @@
+          .sub(/\.erb/, '')
+      end
+
+      app.config.assets.precompile += asset_paths
```
When the gem is added to a Rails app, and you run `rails assets:precompile`, this config allows the gem's assets to be precompiled as separate assets from the app. See https://guides.rubyonrails.org/engines.html#separate-assets-precompiling for more info.

gsmendoza commented on script/install/script.rb:
```diff
@@ -44,26 +72,58 @@ def copy_theme_files
+
+  cp_fontawesome_css_files
+
+  delete_map_files
```
I'm not sure yet how to include the JS map files for precompilation. Since they're not critical, I'm not including them for now.

bigxiang replied:
Rails asset pipeline will generate map files when compiling assets, we don't need the existing map files because they are mapping to the current `min.js` files.

gsmendoza replied:
@bigxiang Hmmm... I'm not sure that's the case here. We're importing the minified JS files from the downloaded Start Bootstrap Admin Theme 2 package. When the Rails asset pipeline compiles the minified JS files, I don't think it will be able to map them to the original JS files. Am I missing something?

bigxiang replied:
Yeah, so my question is why we are importing the minified JS files to Rails asset pipeline. The asset pipeline is using sprocket which uses one kind of JS compressor at back end. It can generate similar result like what other JS build systems do to generate minified JS files. 

So I think it should be good enough to just place the normal JS files into the asset pipeline. If we need to reference to JS files directly, such as from a CDN, it will be the better case to use minified JS files.

What do you think?

gsmendoza replied:
You're right. I chose to include the minified files since they made it easier to automate the import of assets and demo files. 

Currently, I think it should be possible to point the demo files to the normal assets, though I haven't tested this yet.

I'll create an issue to remove the minified files from the gem.

gsmendoza commented on script/install/script.rb:
```diff
@@ -44,26 +72,58 @@ def copy_theme_files
-    updated_contents =
-      contents
-        .gsub('href="vendor/', 'href="/assets/start_bootstrap_admin_theme/vendor/')
-        .gsub('href="css/', 'href="/assets/start_bootstrap_admin_theme/css/')
-        .gsub('src="vendor/', 'src="/assets/start_bootstrap_admin_theme/vendor/')
-        .gsub('src="js/', 'src="/assets/start_bootstrap_admin_theme/js/')
-        .gsub('src="img/', 'src="/assets/start_bootstrap_admin_theme/img/')
+    updated_contents.gsub!(%r{url\(\.\.(.*?)\)}) do
+      %Q{url(<%= asset_path('start_bootstrap_admin_theme/vendor/fontawesome-free/#{$1}') %>)}
+    end
```
It might be better to include Font Awesome's SCSS files instead of changing the paths in the CSS files. Can be done in the future.

gsmendoza reviewed (commented):
@bigxiang Can you review?

@RichardJordan This fixes production-related issues I didn't foresee before with the gem.

gsmendoza committed:
Bump to 0.1.1-alpha

gsmendoza committed:
Fix `gem build` warnings

Remove spec description.
Avoid open-ended dependency on sqlite3.

bigxiang commented on README.md:
```diff
@@ -1,28 +1,39 @@
+
+## Code of Conduct
+
+Everyone interacting in the Kiba::Uncommon project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/SnackNation/start_bootstrap_admin_theme/blob/master/CODE_OF_CONDUCT.md).
```
Not sure what `Kiba::Uncommon` means.

gsmendoza replied:
Copy-pasting fail :) I'll update that. Done.

bigxiang commented on script/install/script.rb:
```diff
@@ -6,6 +6,8 @@
 
 TARGET_VIEWS_PATH = Pathname.new('app/views/start_bootstrap_admin_theme/pages/')
 START_BOOTSTRAP_PATH = Pathname.new(ARGV[0])
+FONTAWESOME_RELATIVE_PATH = 'vendor/fontawesome-free/css'
```
Maybe `font_awesome` is better than `fontawesome` because they are two words: https://github.com/FortAwesome/Font-Awesome?

gsmendoza replied:
@bigxiang Hmmm.. It's a bit more convenient to keep `fontawesome` for the path, since that's the naming used in the theme package download. For convenience, the script retains the directory structure of the theme package when it imports its files to the vendor directory. 

However, I don't mind changing the constant to `FONT_AWESOME_RELATIVE_PATH`. What do you think?

bigxiang replied:
It's not a big deal, just choose the name you feel convenient. :)

bigxiang reviewed (approved):
@gsmendoza We'd better release it first if it works, however I am curious about the intention of this gem. :)

By looking at https://github.com/BlackrockDigital/startbootstrap-sb-admin-2 project, it seems to just add a CSS file and a JS file on the top of bootstrap. Will it be easier to just include them into the admin engine?

Please feel free to merge it.

gsmendoza commented:
@bigxiang 

> By looking at https://github.com/BlackrockDigital/startbootstrap-sb-admin-2 project, it seems to just add a CSS file and a JS file on the top of bootstrap. Will it be easier to just include them into the admin engine?

For the admin engine implementation, the team decided to extract SB Admin 2 theme and its dependencies into a separate gem, so that the dependencies are not managed within the Platform API code.  

When I built this gem, I also had several goals to make its maintenance easier:

1. Write a script to easily import the assets from a SB Admin theme package
2. Incorporate the demo pages so that 
  a. We can confirm how to correctly reference the assets
  b. We know what assets are required per page
  c. The demo can function as a styleguide 

It might be possible to build a more lightweight version of this gem that only has the CSS and JS it needs. However, at this point, I'm still exploring the theme, so I chose to include all its dependencies.

We can discuss this further in the next check in :)

gsmendoza committed:
Fix: Remove KIBA reference

https://github.com/SnackNation/start_bootstrap_admin_theme/pull/2#discussion_r380549169

gsmendoza commented:
Merging this now...
